Add ERC: Solana Storage Router (#390)
ethereum/ERCs: ERCS/erc-7694.md

---
eip: 7694
title: Solana storage handler for CCIP-Write
description: Cross-chain write deferral protocol incorporating storage handler for Solana
author: Avneet Singh (@sshmatrix), 0xc0de4c0ffee (@0xc0de4c0ffee)

> discussions-to: https://ethereum-magicians.org/t/eip-7694-solana-storage-handler-for-ccip-write/19706

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7694-solana-storage-handler-for-ccip-write/19706
```

Author: Fixed in https://github.com/ethereum/ERCs/pull/390/commits/7a09d42beab435c3fc751734c99ec41e367184d7

status: Draft
type: Standards Track
category: ERC
created: 2024-04-18
---

## Abstract
The following standard is an extension to the cross-chain write deferral protocol introducing storage handler for Solana.

## Motivation
[EIP-5559](./eip-5559) introduces two external handlers for deferring write operations to L2s and databases. This document extends that specification by introducing a third storage handler targeting Solana as the storage provider. 

L2s and databases both have centralising catalysts in their stack. For L2s, this centralising agent is the shared security with Ethereum mainnet. In case of databases, the centralising agent is trivial; it is the physical server hosting the database. In light of this, a storage provider that relies on its own independent consensus mechanism is preferred. This specification instructs how the clients should treat write deferrals made to the Solana handler.

> Solana is a cheap L1 solution that is fairly popular among Ethereum community and is widely supported alongside Ethereum by almost all wallet providers. There are several chain-agnostic protocols on Ethereum which could benefit from direct access to Solana blockspace; ENS is one such example where it can serve users of Solana via its chain-agnostic properties while also using Solana's own native storage. This development will encourage more cross-chain functionalities between Ethereum and Solana at core. 

Reviewer: ```suggestion
Solana is a low cost L1 solution that is supported alongside Ethereum by multiple wallet providers. There are several chain-agnostic protocols on Ethereum which could benefit from direct access to Solana blockspace; ENS is one such example where it can serve users of Solana via its chain-agnostic properties while also using Solana's own native storage. This development will encourage more cross-chain functionalities between Ethereum and Solana at core. 
```

Suggest avoiding commentary on popularity with Ethereum community.

---
eip: 7694
title: Solana Storage Router
description: Cross-chain storage router protocol incorporating storage router for Solana
author: Avneet Singh (@sshmatrix), 0xc0de4c0ffee (@0xc0de4c0ffee)
discussions-to: https://ethereum-magicians.org/t/erc-7694-solana-storage-router/19706
status: Draft
type: Standards Track
category: ERC
created: 2024-04-18

> ---

Reviewer 2: ```suggestion
requires: 7700
---
```

If this standard builds on top of 7700, then that proposal is required reading.

## Abstract

> The following standard is an extension to the cross-chain storage router protocol introducing the storage router for Solana blockchain. With this specification, any Ethereum L1 contract can defer a call to Solana blockchain as part of its core functionality, provided that the client is equipped to handle Solana transactions. It was previously possible to defer write and storage operations to other Ethereum L1 contracts, L2 contracts and off-chain databases, and this document extends that functionality to include alternative L1 chains. This standard allows Ethereum to utilise a broader range of cross-chain blockspaces, and puts it well ahead of other L1 ecosystems in its ability to interface with diverse cross-chain infrastructures.

Reviewer 2: Your abstract is missing some technical meat. Summarize&mdash;in the document's abstract&mdash;**how** an Ethereum contract gets access to information on Solana. Does it use an oracle? Does the Ethereum contract contain a Solana light client implementation?

After reading the abstract, I should have a high-level understanding of how the proposal works (and not just what benefits it provides.)

Author: Hi @SamWilsn,

I have added a section describing the methods to read-in Solana data using CCIP-Read. See commit https://github.com/ethereum/ERCs/pull/390/commits/464937ed6b68c79d2a9c241340c86255c6e6a83e

---
eip: 7694
title: Solana Storage Router
description: Cross-chain storage router protocol (CCIP-Store) incorporating storage router for Solana
author: Avneet Singh (@sshmatrix), 0xc0de4c0ffee (@0xc0de4c0ffee)
discussions-to: https://ethereum-magicians.org/t/erc-7694-solana-storage-router/19706
status: Draft
type: Standards Track
category: ERC
created: 2024-04-18
---

## Abstract

> The following standard is an extension to the cross-chain storage router protocol introducing storage router for Solana.

Author: I'll expand on the text and make this document self-sufficient in the next commit.

> The following standard is an extension to the cross-chain storage router protocol introducing storage router for Solana.

Author: I have expanded the abstract as requested in https://github.com/ethereum/ERCs/pull/390/commits/008e6b202f35997a93d9dbca5ae520b02a8ff658

## Motivation

> Cross-Chain Storage Router Protocol, aka CCIP-Store, introduces three external routers for routing storage to L1 contracts, L2s and databases. This document extends that specification by introducing a fourth storage router targeting Solana as the storage provider. 

Reviewer 2: Is CCIP-Store defined in an EIP? If so, you should reference it. If not, please link to it according to the rules in [EIP-1](https://eips.ethereum.org/EIPS/eip-1#linking-to-external-resources).

Author: CCIP-Store is defined in open PR #404 which introduces EIP-7700. I have added the reference to it, although the CI checks will fail until PR #404 is merged and EIP-7700 becomes available.

   | setValue(calldata)                                        |                    |       
   |>>---------------------------->-------------------------->>|                    |       
   |                                                           |                    |       
   | revert StorageRoutedToSolana(programId, account)          |                    |       
   |<<----------------------------<--------------------------<<|                    |       
   |                                                           |                    |       
   | Execute on Solana [programId] [account] [callData]        |                    |       
   |>>---------------------------->---------------------------->---------->------->>|      
   |                                                           |                    |       
   | response                                                  |                    |       
   |<<----------------------------<----------------------------<----------<-------<<|
   |                                                           |                    |       

> ```

Reviewer 2: This is a lovely ascii art diagram, but is there any chance you could provide it as an SVG?

Author: SVG image added in https://github.com/ethereum/ERCs/pull/390/commits/a57b5457833621a5800abaec9d214fb552b1b1a4

## Backwards Compatibility
None.

## Security Considerations
None.

## Copyright

> Copyright and related rights waived via [`CC0`](../LICENSE.md).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

Author: Fixed in https://github.com/ethereum/ERCs/pull/390/commits/a57b5457833621a5800abaec9d214fb552b1b1a4

---
eip: 7694
title: Solana Storage Router
description: Cross-chain storage router protocol incorporating storage router for Solana
author: Avneet Singh (@sshmatrix), 0xc0de4c0ffee (@0xc0de4c0ffee)
discussions-to: https://ethereum-magicians.org/t/erc-7694-solana-storage-router/19706
status: Draft
type: Standards Track
category: ERC
created: 2024-04-18

> requires: 7700

Reviewer 2: ```suggestion
requires: 3668, 7700
```

Author: Fixed ✅
